### PR TITLE
Add the `hashFiles` template func to calculate the SHA256 hash of multiple files

### LINKED ISSUE #196: Support passing multiple files to the `checksum` function
**Is your feature request related to a problem? Please describe.**
While trying to mimick the functionality of the `actions/cache` GitHub Action, I found out that the `checksum` function needed to customize the cache key does not support multiple files

**Describe the solution you'd like**
The input argument of the `checksum` function should be changed so that it accepts multiple strings. It would also be cool if the file paths could be specified as glob

**Describe alternatives you've considered**
A new function with a new name (GitHub calls it `hashFiles`) would be more appropriate and self-explanatory than the current name. The old one could be marked as deprecated and removed in the next major version

### PR DESCRIPTION
**Pull request by ste93cry**
Fixes #196

## Proposed Changes

- Add the `hashFiles` template function

### Description

The `checksum` template function is able to compute the MD5 hash of just one file per time. Instead, similarly to what can be done in the GitHub Actions, it is sometimes useful to be able to get the hash of a series of files. In addition to this, it is also sometimes useful to be able to get the list of files according to a glob pattern. To avoid breaking changes and to also make it clearer that the function can work with a list of files, I've added a new `hashFiles` function.

### Checklist

- [x] Read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] Read the [**CODE OF CONDUCT**](/CODE_OF_CONDUCT.md) document.
- [x] Add tests to cover changes.
- [x] Ensure your code follows the code style of this project.
- [x] Ensure CI and all other PR checks are green
- [x] Add your changes to `Unreleased` section of [CHANGELOG](/CHANGELOG.md).
- [x] Ensure [documentation](/DOCS.md) is up-to-date. The same file will be updated in [plugin index](https://github.com/drone/drone-plugin-index/blob/master/content/meltwater/drone-cache/index.md) when your PR is accepted, so it will be available for end-users at http://plugins.drone.io.

### CODE REVIEW

File: key/generator/metadata.go [hunk 1]
```diff
@@ -80,26 +83,63 @@ func (g *Metadata) parseTemplate() (*template.Template, error) {
...
+
+	f, err := os.Open(path)
+	if err != nil {
+		level.Error(logger).Log("cache key template/checksum could not open file")
+		return []byte{}
+	}
+
+	defer internal.CloseWithErrLogf(logger, f, "checksum close defer")
+
+	str, err := readerHasher(f)
```
Comment (kakkoyun) [inline key/generator/metadata.go:138]:
```suggestion
	h, err := readerHasher(f)
```

nit

File: key/generator/metadata.go [hunk 2]
```diff
@@ -80,26 +83,63 @@ func (g *Metadata) parseTemplate() (*template.Template, error) {
...
+			level.Error(logger).Log("could not generate the hash of the input files: %s", err.Error())
 			return ""
 		}
 
-		return str
+		return fmt.Sprintf("%x", h)
+	}
+}
+
+func getFileHash(path string, logger log.Logger) []byte {
```
Comment (kakkoyun) [inline key/generator/metadata.go:123]:
```suggestion
func fileHash(path string, logger log.Logger) []byte {
```

**Review by kakkoyun — APPROVED** [review]
Amazing work :100: Thanks a lot :rocket: 

I have just commented a couple of nit-pickings. Feel free to ignore them!

I don't have the merge rights any more, so I'm leaving the decision to merge to @meltwater-foundation people

**Review by kakkoyun — DISMISSED** [review]

**Comment by ste93cry** [comment]
Build failure is unrelated, if some member of @meltwater-foundation can restart the CI it would be 🆒

**Review by kakkoyun — DISMISSED** [review]

**Comment by KucharczykL** [comment]
Hi, I've found this PR when looking for docs about the `checksum` function to see if it supports multiple files and globs. It would be really nice to have this!

**Comment by ste93cry** [comment]
@hikerspath can you 🙏 review this?

**Comment by ste93cry** [comment]
@cydowens can you please review this?

**Comment by bdebyl** [comment]
@ste93cry Do you mind resolving the merge conflicts by pulling latest master into your working branch?

**Comment by ste93cry** [comment]
While we're at it, I was wondering if it makes sense to change the hashing algorithm to SHA256 like GitHub. WDYT?

**Comment by bdebyl** [comment]
@ste93cry I'll leave this entirely up to you. As it's just simple file hashes I don't see a problem sticking with md5, but I fully support either! Let me know if you want me to create a separate issue to change from md5 to sha2 or if you'd like to just implement the change to use sha2 with this PR.

Again, either way is fine by me and thanks for the good work :)

**Comment by ste93cry** [comment]
>As it's just simple file hashes I don't see a problem sticking with md5

The main problem is that you may get the same hash for different files. More details can be found [here](https://en.wikipedia.org/wiki/MD5#Collision_vulnerabilities)

>Let me know if you want me to create a separate issue to change from md5 to sha2 or if you'd like to just implement the change to use sha2 with this PR

Let's implement the changes here, they are minimal anyway so it won't take much more time to implement them

**Comment by ste93cry** [comment]
>Let me know if you want me to work on it, or if you'll be taking the charge on the code changes necessary for this change

Changes incoming, don't worry 😃

**Comment by ste93cry** [comment]
@bdebyl changes are ready, do you mind reviewing them?

**Review by bdebyl — APPROVED** [review]

**Comment by bdebyl** [comment]
@ste93cry sorry for getting back to you so late, it looks good to me pending one more engineering team member review and then we can merge and stage it for v1.4.0!

**Review by hikerspath — APPROVED** [review]

**Pull request merged**